### DISCUSSION TITLE
remove Hydrogens on a molecule but did not obtain an expected molecule

### DISCUSSION THREAD

**Discussion by tsn2027** [rdkit/rdkit#6613] [Q&A] [answered]
Hello all,

I am trying to remove Hydrogens on a molecule (shown in a snippet below). 

When I used ```Chem.RemoveHs()```, it did not remove any Hs unless I used ```Chem.RemoveAllHs()```. In case I used ```Chem.RemoveAllHs()```, I didn't obtain the molecule that I am expecting. I also tried to use ```Chem.rdmolops.RemoveHsParameters()``` to remove Hs and get an expected molecule, but it was not successful.

May I ask whether there is a way to remove Hs and still be able to match an expected molecule? 

Thanks in advances.

```
sample_smiles = "[1H]C([2H])([3H])C([4H])([5H])C([6H])([7H])C([8H])([9H])C([10H])([11H])" # C5H11
molRemoveAllHs = Chem.RemoveAllHs(Chem.MolFromSmiles(sample_smiles))
print("molRemoveHs:",Chem.MolToSmiles(molRemoveAllHs))

params = Chem.rdmolops.RemoveHsParameters()
params.removeIsotopes = True
# params.removeMapped = True
mol = Chem.RemoveHs(Chem.MolFromSmiles(sample_smiles), params)

display(molRemoveAllHs)
display(mol)
print()

print("Expected molecule:")
display(Chem.MolFromSmiles("CCCC[CH2]"))
```

```
rdkit version:  2023.03.2
python version: 3.10.12
```

**Comment by bp-kelley** [comment]
If you declare an isotope on an H, you can’t remove hs.  Rdkit (and smiles) only supports implicit hydrogens in their normal state.

**Reply by tsn2027** [reply]
Just wondering if there is a way to convert a current smiles with isotope H to an expected molecule by chance?

**Reply by tsn2027** [reply]
What I mean is if there is a direct function in rdkit to do that? 

I am thinking is to write a custom function to obtain an expect molecule, but I am afraid that there is a defined function rdkit to do the job for me and I missed it out.

**Comment by ptosco** [comment, marked as answer]
@tungnguyen2027 By default, RDKit will fill all open valences which bear no hydrogen atoms with implicit Hs.
What you need here is to restore your radical molecule by converting the implicit H into a radical electron, and then remove all isotopic Hs:
```
from rdkit import Chem

sample_smiles = "[1H]C([2H])([3H])C([4H])([5H])C([6H])([7H])C([8H])([9H])C([10H])([11H])" # C5H11
mol = Chem.MolFromSmiles(sample_smiles)

for a in mol.GetAtoms():
    if a.GetNumImplicitHs():
        a.SetNumRadicalElectrons(a.GetNumImplicitHs())
        a.SetNoImplicit(True)
        a.UpdatePropertyCache()
molRemoveAllHs = Chem.RemoveAllHs(mol)

Chem.MolToSmiles(molRemoveAllHs)
'[CH2]CCCC'

molRemoveAllHs
```
![image](https://github.com/rdkit/rdkit/assets/5244385/5564f1db-4402-4e14-96b8-c6949e33667c)

**Reply by tsn2027** [reply]
Thank you all so much for the help, @ptosco @bp-kelley! I am closing this discussion.

**Reply by tsn2027** [reply]
Following up with this question, when I executed this function for this smiles ```[O]O```, it gives me this ```[O][O]``` instead, and it is incorrect. May I ask if there is a way to apply this function to any smiles which has isotope H, @ptosco ?

**Reply by bp-kelley** [reply]
You have two different examples here.   It looks like you are assuming that all explicit hydrogens in the graph exist as `[H]` in the smiles string.

For example

```
C([H])([H])
```

you want to be `[CH2]`

As an aside, If we parse this using normal smiles rules, you would get `[CH4]`, to make it CH2 you would need to have

```
[CH0]([H])([H])
```

or (more canonically)

```
[CH2]
```

Now the code @ptosco sent assumes that every hydrogen explicitly exists in the graph, i.e. `[H]` so the code will give `[CH2]` for your example, ***but this is directly rejecting the original smiles rules for parsing hydrogens***

For the second example, you don't specify hydrogens:

```
[O]O
```

and this, correctly under the assumptions that the code makes, yields to oxygen with no hydrogens

```
[O][O]
```

For @ptosco's code (and as seen in your first example) you would need to have given:

```
[O]O([H])
```


Might I ask where are these structures coming from?  I think this rabbit hole you are in requires quite a few assumptions and different ones for the examples you have given.

**Reply by tsn2027** [reply]
Thank you for your detail explanation. 

These structures are generated by me. What I am trying to do is that for any structures that have isotope H, I want to have @ptosco's code applied to convert them to the cano form; otherwise, it does not apply.
